### DISCUSSION TITLE
Mechanism for working with vision-based models

### DISCUSSION THREAD

**Discussion by daedsidog** [karthink/gptel#245] [Ideas]
A lot of times I need to pass information to ChatGPT that I can't copy, such as a snippet from an old, scanned document or formatted mathematics.

Right now I have a manual process where I query a visual model (though a website) so that he tells me what he sees. E.g., when I paste him a snippet of mathematics, he will give me LaTeX code, which I then pass down to ChatGPT.

Would be very nice to have something like this.

**Comment by karthink** [comment]
I forget, do you use org-mode or markdown with gptel?

**Comment by daedsidog** [comment]
I used to use Org mode but I switched to Markdown because I was tired of gptel sometimes doing weird things (like removing underscores). I think it's fixed in the latest version, but I haven't switched back.

Why is that relevant, though?

**Comment by karthink** [comment]
> I used to use Org mode but I switched to Markdown because I was tired of gptel sometimes doing weird things (like removing underscores). I think it's fixed in the latest version, but I haven't switched back.

It should be fixed now, yeah.

> Why is that relevant, though?

It's easier to support vision models in Org mode.  That said, please see the discussion in #231.

**Comment by daedsidog** [comment]
> It's easier to support vision models in Org mode.  That said, please see the discussion in [#231](https://github.com/karthink/gptel/discussions/231).

Interesting.

I honestly think the most power from gptel comes with just the abstraction layer it provides when interacting with various models. I, for one, have completely eliminated the process of manually typing code by implementing context generation. Below is a demonstration of me constructing a context buffer, and with a keypress I use gptel's replace-in-place with it. Works exceedingly well. This is kind of its own separate thing from gptel, but I was wondering if the scope of gptel should include this sort of thing.

![contexter](https://github.com/karthink/gptel/assets/41439659/aca82b11-2b97-46e1-88a5-177653316059)

What I want now is just a way, totally unrelated to Org mode or MD, which will allow me to "send" ChatGPT queries with images (i.e., send current image saved on clip) and get an input in place.

**Comment by karthink** [comment]
> I honestly think the most power from gptel comes with just the abstraction layer it provides when interacting with various models

I see.

> Below is a demonstration

Sorry, I had trouble following your demo.  My best guess is that the buffer on the right is sent as the context (or system message), and you're asking it to do something with those functions.  

- I don't understand how it knows what "special function" means, or what "process" means,
- or what the utility of this approach is in more general situations.

> What I want now is just a way, totally unrelated to Org mode or MD, which will allow me to "send" ChatGPT queries with images (i.e., send current image saved on clip) and get an input in place.

I'm not sure gptel is set up to do that -- it's a very buffer-oriented system.  At minimum it will need to distinguish between text as text and text that represents a file path and act on the file instead.  A common way to do this would be to define a `gptel-send-image` command, but I'm not interested in growing the command surface area of gptel.

Basically, handling images is not ruled out, but right now I don't know the best way of doing so that conforms to a simple mental model like the chat usage does.

**Comment by karthink** [comment]
>  I was wondering if the scope of gptel should include this sort of thing.

I'm interested to understand what you mean here -- I just had trouble following the demo.

**Comment by daedsidog** [comment]
> > I was wondering if the scope of gptel should include this sort of thing.
> 
> I'm interested to understand what you mean here -- I just had trouble following the demo.

My apologies, my explanation was terrible.

The demo showcases a way to mark areas in different buffers, and aggregate them in their own dedicated buffer. That buffer can then be copied and handed to gptel as context. This is much easier than manually copy pasting sections of context into the dedicated chat buffer/external ChatGPT website, and also has the added bonus of minimizing the context by collapsing code that doesn't contribute to the context.

You can manually remove context snippets from the context buffer.

In a nutshell, it's a glorified yanker, but I found it incredibly useful.

> I'm interested to understand what you mean here

I am wondering if you would be open for this to be integrated into gptel, or should this remain its own separate package. It's pretty useless outside of gptel, though.

**Comment by karthink** [comment]
> I am wondering if you would be open for this to be integrated into gptel, or should this remain its own separate package. It's pretty useless outside of gptel, though.

I like the idea!  I'll have to think about how to integrate it into gptel though.  Right now the best idea I have is "Add an option to the transient menu to append a selected region to the system prompt".  This won't work well across buffers since each buffer has its own system prompt.

You've developed a more sophisticated UI for this style of usage, it's interesting.

**Comment by karthink** [comment]
Converting to a discussion since there's nothing to fix in gptel right now.

**Comment by daedsidog** [comment]
> > I am wondering if you would be open for this to be integrated into gptel, or should this remain its own separate package. It's pretty useless outside of gptel, though.
> 
> I like the idea! I'll have to think about how to integrate it into gptel though. Right now the best idea I have is "Add an option to the transient menu to append a selected region to the system prompt". This won't work well across buffers since each buffer has its own system prompt.
> 
> You've developed a more sophisticated UI for this style of usage, it's interesting.

Just mind you, it's not a system prompt. I lexically set a system prompt that tells it how to treat the text it's supposed to replace, then I insert the user prompt in place with what is supposed to be replaced, and then I use the gptel refactor to handle everything.

**Comment by doctorguile** [comment]
@daedsidog your demo is very interesting

If you don't mind, can share your gptel add-on so we can try it out?

Thanks

**Comment by daedsidog** [comment]
@doctorguile I'll add all the things to my fork sometime soon.

**Comment by karthink** [comment]
I have added vision support to gptel in the `feature-capabilities` branch.  Currently all backends except Gemini are supported.  This is not a proof of concept hack like my previous attempt, I did it the "right" (and painful) way this time -- please test if you're interested.  It's a pretty big change so there are sure to be bugs.

It's actually a little more general than vision support -- a lot of the changes are about specifying per-model capabilities, to pave the way to add function calling, JSON output and image output (DALL-E etc) uniformly to `gptel-request`.

To set it up correctly,
1. ensure that your `gptel-model` is a symbol, not a string (i.e. `gpt-4o-mini` instead of `"gpt-4o-mini"`), and set to a model that supports vision (except Gemini models).
2. See the documentation of `gptel-make-*`, where `*` is `openai`, `anthropic` or `ollama`,
3. and the documentation of the new user option `gptel-track-media`.

There are two ways to use it.
1. Set `gptel-track-media` to `t` and type in a link to an image file, on a line by itself, in an Org or Markdown chat buffer.  Then `gptel-send` will find and send the image along with your text.
2. Add an image file to the context using `gptel-add-file`, or `gptel-add` from an `image-mode` buffer.

**Comment by wlauppe** [comment]
Hello Karthik what wonderful news! I was hoping for a long time you would implement vision capabilites. 
Here are my test results: 
## Method 1. 
Using this method works perfectly. 
I added the image as context via the ```gptel-menu``` -f option
asked "Describe the picture" 
and got back a satisfactory answer.

```The image shows a simple, line-drawn outline of a house. The house has a basic structure with a peaked roof on the left side and a flat roof extension on the right. There's a rectangular door in the center and two square windows with cross-frames on the right side. A chimney is visible on the roof. The drawing is entirely in black and white, with no shading or detail beyond the basic outlines. This style of drawing is often used in coloring books or as a simple representation of a home in diagrams or children's illustrations.```
## Method 2: 
i marked the region and pressed my ```gptel-send```- shortcut, but was not successful. 

![screenshot_20240930_165321](https://github.com/user-attachments/assets/cd45e338-9ec7-492c-ae07-aa0d4668351f)
I read the documentation of  the ```gptel-track-media``` variable
and the readme of the branch, there you wrote, that at the moment this method is only supported in chat mode. But in chat mode it didnt work either. 
my image link is of this form. 
```
[[file:orgdownload/20240930_083921_screenshot_20240930_083918.png]]
```
it conforms. to the org mode image link defaults.  
```An image is a link to an image file that does not have a description part``` 
https://orgmode.org/manual/Images.html

But this doesnt work. 
in your documentation you say, it has to be a different format. Can you give an image link example that works?

**Reply by karthink** [reply]
Any of the following should work:

```
Describe the picture

[[file:/tmp/screenshot_20240930T160200.png]]
```

```
Describe the picture

[[file:/tmp/screenshot_20240930T160200.png][some picture]]
```

```
Describe the picture

<file:/tmp/screenshot_20240930T160200.png>
```

I just tested them.  Note that the link is on a line by itself, with no other text around it.

This will not work:
```
Describe the picture

file:/tmp/screenshot_20240930T160200.png
```

----
To test it, you don't have to send the query.  You can run `(setq gptel-expert-commands t)`, then pick the dry-run option from the transient menu.  That will show you what data will be sent.  (The image should be encoded as a base64 string.)

I'm also interested in your intuition about what link format gptel should and shouldn't send -- but this is a different issue.  First let's confirm that the above is working as intended.

**Reply by wlauppe** [reply]
I cant get eithher of your 3 "should-work-links" to work. 
Did you get them to work even without chatmode?
Did you get them to work in org mode, or only in markdown mode?
my system settings are:
```
Emacs 29.4, linux os
```
marking a region and pressing ```gptel-send``` returns 

```Wrong type argument: listp, paragraph```

for switching from master to the ```feature-capabilities``` branch I did the following. 
1.
```
cd ~/.emacs.d/straight/repos/gptel 
git checkout feature-capabilities
git pull 
```
2. in emacs
```
M-x straight-rebuild-package
gptel
```
3. 
restarted emacs. 

Is this the way you would do it, or should i do it differently?

**Reply by karthink** [reply]
> Did you get them to work even without chatmode?

Links are recognized and sent as images in chat buffers only (i.e. with `gptel-mode` on).

> Did you get them to work in org mode, or only in markdown mode?

Both.

> I cant get eithher of your 3 "should-work-links" to work.

Have you set `gptel-track-media` to `t`?  You can do this via elisp, OR via the button in the header-line.

![image](https://github.com/user-attachments/assets/8bd91ef1-b85f-461f-84ce-25590aaecb8f)

If you don't see this button it means the current model doesn't support images, or gptel thinks it doesn't.

**Reply by wlauppe** [reply]
![gptel_vision](https://github.com/user-attachments/assets/cd52a1dd-d98e-4c55-bb7e-dde4f8efb899)

hello Karthik, yes I did. I setup a new virtual machine with plain vanilla emacs 29.4 and only the straight package manager, markdown, and gptel enabled, just to rule out, that my configuration plays a role. It seems to work in markdown mode, it doesnt work in org-mode. 

the buffer "test.md" shown in the upper half, after pressing gptel-send i get the response shown. It clearly recognizes the image. In the "test.org" buffer shown below, I get the error shown in in the screenshot in the mode line:

**Wrong type argumetn: listp, paragraph**
Do you have any idea how to debug this further?

**Reply by wlauppe** [reply]
he image was this one from wikipedia, converted to png format.
https://de.wiktionary.org/wiki/cat#/media/Datei:Cat03.jpg

**Reply by karthink** [reply]
Couple of things:

The link has to be isolated from text, so this won't work:
```
Describe the following picture
[[./test.png][test]]

```
whereas this should
```
Describe the following picture

[[./test.png][test]]
```

Independent of this, the `wrong type argument` error needs to be looked into.  Can you produce a backtrace by running `toggle-debug-on-error` first?

**Reply by karthink** [reply]
The reason for the isolated link criterion is that we don't want to send inline links like this:

```
In [[./figures/ref.png][figure 3]], we consider ...
```

The link-type criterion for sending can be made user-customizable, but I need to verify that the branch works as intended right now so I can merge it into the master branch.

**Reply by wlauppe** [reply]
it doesnt work, even with newlines before and after the link
![screenshot_20241007_212225](https://github.com/user-attachments/assets/fe992f19-b9b1-408e-aaa7-d1ebb2a4785e)

i attached the backtrace, with the key replaced with : key "sk-xxx"
[backtrace.txt](https://github.com/user-attachments/files/17283549/backtrace.txt)

**Reply by karthink** [reply]
I've pushed a potential fix, please test.  (The issue was an API change in the development version of Org.)

**Reply by wlauppe** [reply]
running your latest version i get a new error: 
```gptel-org--link-standalone-p: Symbol’s function definition is void: org-element-begin```

**Reply by wlauppe** [reply]
thats the backtrace 
[backtrace2.txt](https://github.com/user-attachments/files/17289261/backtrace2.txt)

**Reply by wlauppe** [reply]
i rebuild emacs using the master branch of 
```git://git.sv.gnu.org/emacs.git``` ```emacs --version``` returns ```GNU Emacs 31.0.50```

there everything works perfectly, but im happy to test gptel using the default ubuntu emacs 29.4 version as long as you need it.

**Reply by karthink** [reply]
> running your latest version i get a new error: `gptel-org--link-standalone-p: Symbol’s function definition is void: org-element-begin`

You're on a very old version of Org.  Anyway, I pushed a fix working around this issue.  It should now work in the default Ubuntu Emacs 29.4 version.

**Reply by wlauppe** [reply]
your latest version seems to work also with emacs version 29.4. that seems to be fixed, now. Great 👍

**Comment by karthink** [comment]
Image support has been merged, and is available in gptel 0.9.5.

**Comment by daedsidog** [comment]
There is no support for Markdown mode yet, right? I tried multiple link variations and none of them were detected. Looking in the code, only org mode has link parsing.

**Reply by karthink** [reply]
Image or document links in markdown are supported.  The documentation for `gptel-track-media` explains.

Plus you can add a media file to the context with `gptel-add` irrespective of major-mode.

**Reply by daedsidog** [reply]
I don't know why it did not occur to me to read its documentation.

Nevertheless, the angled bracket format does not seem to work still, while the other format (i.e. the `[name](link/to/media)`) does:

![image](https://github.com/user-attachments/assets/a53936c5-09a2-4e3a-980a-715e0f753b8b)

![image](https://github.com/user-attachments/assets/d6e98cd5-68ce-4f7b-87c5-0baa78cb4c22)

Incidentally, this was the format I tried, which led me to my assumptions.

This in my opinion is actually a very nice integration of vision. Great work.

**Reply by karthink** [reply]
Angle bracket links in markdown syntax need to have the full URI to be recognized as a link. i.e. you need

```markdown
<file:///tmp/clipboard52.png>

What do you see
```

**Reply by daedsidog** [reply]
Thanks, that worked. I tried `<file:/tmp/etc.png>` before, like in org, and that's why it didn't work.

Cheers!
